### PR TITLE
feat(partition): move large state table to a partitioned compaction g…

### PR DESCRIPTION
**Pull request by soundOfDestiny**
…roup

I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

move large state table to a partitioned compaction group

## Checklist For Contributors

- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Review by Little-Wallace — APPROVED** [review]
0 represents not partition, does it? Please add comment for this.
rest LGTM

File: src/meta/src/hummock/manager/compaction_group_manager.rs [hunk 1]
```diff
@@ -428,7 +429,8 @@ impl<S: MetaStore> HummockManager<S> {
         parent_group_id: CompactionGroupId,
         table_ids: &[StateTableId],
     ) -> Result<CompactionGroupId> {
-        self.move_state_table_to_compaction_group(parent_group_id, table_ids, None, false)
+        // 0 represents no vnode partition
+        self.move_state_table_to_compaction_group(parent_group_id, table_ids, None, false, 0)
```
Comment (Li0k) [inline src/meta/src/hummock/manager/compaction_group_manager.rs:433]: Is it better to use Option than magic number "0" ?

**Review by Li0k — APPROVED** [review]
Rest LGTM, look forward to the test result, thanks for the effort

File: src/meta/src/hummock/manager/compaction_group_manager.rs (hunk 1 above)
Comment (soundOfDestiny) [reply src/meta/src/hummock/manager/compaction_group_manager.rs:433]: > Is it better to use Option than magic number "0" ?

It's hard to use `Option` in protobuf. :lark-cry:

Comment (Li0k) [reply src/meta/src/hummock/manager/compaction_group_manager.rs:433]:
```rust
pub async fn move_state_table_to_compaction_group(
        &self,
        parent_group_id: CompactionGroupId,
        table_ids: &[StateTableId],
        target_group_id: Option<CompactionGroupId>,
        allow_split_by_table: bool,
        split_weight_by_vnode: Option<u32>,
    )
```

like this ?

Comment (soundOfDestiny) [reply src/meta/src/hummock/manager/compaction_group_manager.rs:433]: > ```rust
> pub async fn move_state_table_to_compaction_group(
>         &self,
>         parent_group_id: CompactionGroupId,
>         table_ids: &[StateTableId],
>         target_group_id: Option<CompactionGroupId>,
>         allow_split_by_table: bool,
>         split_weight_by_vnode: Option<u32>,
>     )
> ```
> 
> like this ?

Yes but we cannot store `Option` in protobuf.
From another point of view, compaction needs to cut the state table every time `1/weight` of vnodes in the table have been processed. Therefore when `weight == 0`, `1/weight == inf` so we will never cut.

**Comment by soundOfDestiny** [comment]
> Rest LGTM, look forward to the test result, thanks for the effort

There isn't any good result. Perhaps the strategy of the compaction picker is still the old strategy.
How can we enable the new strategy in compaction picker?

**Comment by Li0k** [comment]
> > Rest LGTM, look forward to the test result, thanks for the effort
> 
> There isn't any good result. Perhaps the strategy of the compaction picker is still the old strategy. How can we enable the new strategy in compaction picker?

Enable by default

**Comment by soundOfDestiny** [comment]
nexmark q7 benchmark:
PR: 638208.122222
main: 624712.056111

**Comment by soundOfDestiny** [comment]
nexmark q4+q7 benchmark:
PR:
compactor CPU:
截屏2023-05-16 14 14 20

**Comment by soundOfDestiny** [comment]
After q4, q7 and q20 benchmark tests, we do not find any improvement.
cc @Little-Wallace

**Comment by soundOfDestiny** [comment]
nexmark q20 benchmark result:
source throughput:
PR(split=4):
截屏2023-05-17 14 29 18

main:
截屏2023-05-17 14 29 50

**Comment by Li0k** [comment]
At the q4 + q7 test case

image

We can observe that the flush traffic of cn is relatively low, and we may not observe the optimization of split by vnode when the pressure of to base compaction is low.

**Comment by Li0k** [comment]
> nexmark q20 benchmark result: source throughput: PR(split=4): 截屏2023-05-17 14 29 18
> 
> main: 截屏2023-05-17 14 29 50

https://g-2927a1b4d9.grafana-workspace.us-east-1.amazonaws.com/d/EpkBw5W4k/risingwave-test-dashboard?orgId=1&from=1683859130000&to=1683860936000&var-namespace=nexmark-blackhole-q20

image

I found some problems with grafana in branch testing, where the compactor is in an abnormal state, which may lead to inaccurate test conclusions

**Comment by soundOfDestiny** [comment]
> > nexmark q20 benchmark result: source throughput: PR(split=4): 截屏2023-05-17 14 29 18
> > main: 截屏2023-05-17 14 29 50
> 
> https://g-2927a1b4d9.grafana-workspace.us-east-1.amazonaws.com/d/EpkBw5W4k/risingwave-test-dashboard?orgId=1&from=1683859130000&to=1683860936000&var-namespace=nexmark-blackhole-q20
> 
> image
> I found some problems with grafana in branch testing, where the compactor is in an abnormal state, which may lead to inaccurate test conclusions

Good catch. I will make another try.

**Comment by hzxa21** [comment]
> I found some problems with grafana in branch testing, where the compactor is in an abnormal state, which may lead to inaccurate test conclusions

It seems that compactor keeps dying due to OOM. Is it due to some bugs/misconfigs? We should confirm that before making another try. Otherwise the same issue may happen again.

**Comment by soundOfDestiny** [comment]
> > At the q4 + q7 test case
> > image
> > We can observe that the flush traffic of cn is relatively low, and we may not observe the optimization of split by vnode when the pressure of to base compaction is low.
> 
> In your experiments which are mentioned in #8707, the flush traffic of cn is also relatively low. 截屏2023-05-17 15 37 25

However, the new scheme still shows better performance in spite of the low flush traffic in your experiments.

**Comment by soundOfDestiny** [comment]
I have done the retry for PR(split=4), and there is no OOM in retry.
However, it still shows performance regression comparing to main:

nexmark q20:
grafana:
https://g-2927a1b4d9.grafana-workspace.us-east-1.amazonaws.com/d/EpkBw5W4k/risingwave-test-dashboard?orgId=1&from=1684323410000&to=1684325214000&var-namespace=nexmark-blackhole-q20

source throughput:
截屏2023-05-17 20 15 03

**Comment by Li0k** [comment]
> 

Thanks for the test,

IMO, the compactor is in an abnormal state.

image

image

image

i guess some stuck happening due to a long running task,  but i'm not quite sure. Any other ideas? @hzxa21 @Little-Wallace 
This phenomenon is accompanied by the occurrence of split. I think this is still related to our picker strategy. The alignment was not done before the split, and a large task was selected.

my suggestion is
1. Try to trigger the split at an earlier stage

2. Test different split granularity

**Comment by soundOfDestiny** [comment]
> > 
> 
> Thanks for the test,
> 
> IMO, the compactor is in an abnormal state.
> 
> image image image
> i guess some stuck happening due to a long running task, but i'm not quite sure. Any other ideas? @hzxa21 @Little-Wallace This phenomenon is accompanied by the occurrence of split. I think this is still related to our picker strategy. The alignment was not done before the split, and a large task was selected.
> 
> my suggestion is
> 
> 1. Try to trigger the split at an earlier stage
> 2. Test different split granularity

I have also tested `split=2`.

**Comment by soundOfDestiny** [comment]
> > > > 
> > > 
> > > 
> > > Thanks for the test,
> > > IMO, the compactor is in an abnormal state.
> > > image image image
> > > i guess some stuck happening due to a long running task, but i'm not quite sure. Any other ideas? @hzxa21 @Little-Wallace This phenomenon is accompanied by the occurrence of split. I think this is still related to our picker strategy. The alignment was not done before the split, and a large task was selected.
> > > my suggestion is
> > > 
> > > 1. Try to trigger the split at an earlier stage
> > > 2. Test different split granularity
> > 
> > 
> > I have also tested `split=2`.
> 
> Thanks for the test, Is the compactor cpu resources not being fully utilized due to the introduction of partitioning? Or is there some other reason for the abnormal cpu and memory utilization in the test?

I found that in the report in https://github.com/risingwavelabs/risingwave/issues/8707, you mentioned that
>
> 1. try to break the assumption of sub_level compaction and increase the parallelism of sub_lvel compaction by picking multiple sub_level files at once instead of picking the whole level.
> 2. split the l0 and base level files into specified partitions by means of vnode and so on, thus improving the efficiency of 1.
> 3. When there are multiple state tables in the system, move the state table data with more writes to a separate compaction group to enable the above optimization and thus improve the system throughput.
> 
> I believe that the above optimizations are implemented independently of each other, but they can be used in conjunction with each other, and I have verified this through the above tests.
> 
Is this possible to reproduce the result you mentioned on main?

**Pull request closed without merging**